### ISSUE TITLE
Pytorch 2.4.0 crashes with illegal instruction error on Raspberry Pi

### ISSUE THREAD

**Issue by eltimen** [pytorch/pytorch#132032]
### 🐛 Describe the bug

When I try to import latest torch 2.4.0, on my Raspberry Pi 4 board, `python3` crashes with Illegal instruction error, but the previous 2.3.1 version works fine.
Steps to reproduce:
```
pip install torch==2.4.0
python3
import torch
```
The SIGILL is caused by `ldaddal` instruction inside `torch.libs/libarm_compute-84b761c0.so`. This instruction is part of LSE and isn't available on my Cortex-A72 processor.
I have found the similar ticket is #97226, but it is for old 2.0.0 version
Details:
```
(test_torch) user@raspberrypi:~ $ gdb python3
(gdb) start
Function "main" not defined.
Make breakpoint pending on future shared library load? (y or [n])
Starting program: /home/user/test_torch/bin/python3
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch

Program received signal SIGILL, Illegal instruction.
0x0000007feb8c1450 in std::string::_Rep::_M_dispose(std::allocator<char> const&) [clone .part.0] ()
   from /home/user/test_torch/lib/python3.9/site-packages/torch/lib/../../torch.libs/libarm_compute-84b761c0.so
(gdb) disassemble
Dump of assembler code for function _ZNSs4_Rep10_M_disposeERKSaIcE.part.0:
   0x0000007feb8c1424 <+0>:     adrp    x2, 0x7febad1000 <_ZTVN8arm_conv8winograd15input_transform17TransformUnpaddedIDhDhEE+24>
   0x0000007feb8c1428 <+4>:     ldr     x2, [x2, #3392]
   0x0000007feb8c142c <+8>:     cbnz    x2, 0x7feb8c1448 <_ZNSs4_Rep10_M_disposeERKSaIcE.part.0+36>
   0x0000007feb8c1430 <+12>:    ldr     w2, [x0, #16]
   0x0000007feb8c1434 <+16>:    sub     w3, w2, #0x1
   0x0000007feb8c1438 <+20>:    str     w3, [x0, #16]
   0x0000007feb8c143c <+24>:    cmp     w2, #0x0
   0x0000007feb8c1440 <+28>:    b.le    0x7feb8c1458 <_ZNSs4_Rep10_M_disposeERKSaIcE.part.0+52>
   0x0000007feb8c1444 <+32>:    ret
   0x0000007feb8c1448 <+36>:    add     x3, x0, #0x10
   0x0000007feb8c144c <+40>:    mov     w2, #0xffffffff                 // #-1
=> 0x0000007feb8c1450 <+44>:    ldaddal w2, w2, [x3]
   0x0000007feb8c1454 <+48>:    b       0x7feb8c143c <_ZNSs4_Rep10_M_disposeERKSaIcE.part.0+24>
   0x0000007feb8c1458 <+52>:    b       0x7feb2f1610 <_ZNSs4_Rep10_M_destroyERKSaIcE@plt>
End of assembler dump.
(gdb) bt
#0  0x0000007feb8c1450 in std::string::_Rep::_M_dispose(std::allocator<char> const&) [clone .part.0] ()
   from /home/user/test_torch/lib/python3.9/site-packages/torch/lib/../../torch.libs/libarm_compute-84b761c0.so
#1  0x0000007feb30f7e0 in _GLOBAL__sub_I_weight_transforms_fp16.cpp () from /home/user/test_torch/lib/python3.9/site-packages/torch/lib/../../torch.libs/libarm_compute-84b761c0.so
#2  0x0000007ff7fdac9c in call_init (l=<optimized out>, argc=argc@entry=1, argv=argv@entry=0x7ffffff4a8, env=env@entry=0x7ffffff4b8) at dl-init.c:72
#3  0x0000007ff7fdada4 in call_init (env=0x7ffffff4b8, argv=0x7ffffff4a8, argc=1, l=<optimized out>) at dl-init.c:30
#4  _dl_init (main_map=0xb2c850, argc=1, argv=0x7ffffff4a8, env=0x7ffffff4b8) at dl-init.c:119
#5  0x0000007ff7ddc144 in __GI__dl_catch_exception (exception=0x0, operate=0x7ff7fde130 <call_dl_init>, args=0x7fffffc3c0) at dl-error-skeleton.c:182
#6  0x0000007ff7fded28 in dl_open_worker (a=a@entry=0x7fffffc610) at dl-open.c:758
#7  0x0000007ff7ddc0e4 in __GI__dl_catch_exception (exception=0x7fffffc5f8, operate=0x7ff7fde9b0 <dl_open_worker>, args=0x7fffffc610) at dl-error-skeleton.c:208
#8  0x0000007ff7fde610 in _dl_open (file=0x7ff73661d0 "/home/user/test_torch/lib/python3.9/site-packages/torch/_C.cpython-39-aarch64-linux-gnu.so", mode=-2147483646,
    caller_dlopen=0x5ca9d0, nsid=-2, argc=1, argv=0x7ffffff4a8, env=0x7ffffff4b8) at dl-open.c:837
#9  0x0000007ff7f670b4 in dlopen_doit (a=a@entry=0x7fffffc8c8) at dlopen.c:66
#10 0x0000007ff7ddc0e4 in __GI__dl_catch_exception (exception=exception@entry=0x7fffffc850, operate=0x7ff7f67050 <dlopen_doit>, args=0x7fffffc8c8) at dl-error-skeleton.c:208
#11 0x0000007ff7ddc1b0 in __GI__dl_catch_error (objname=0x943a00, errstring=0x943a08, mallocedp=0x9439f8, operate=<optimized out>, args=<optimized out>) at dl-error-skeleton.c:227
#12 0x0000007ff7f67960 in _dlerror_run (operate=operate@entry=0x7ff7f67050 <dlopen_doit>, args=args@entry=0x7fffffc8c8) at dlerror.c:170
#13 0x0000007ff7f67154 in __dlopen (file=<optimized out>, mode=<optimized out>) at dlopen.c:87
```

### Versions

The `collect_env.py` crashes, so:
* OS: Debian 11
* Python 3.9.2
* lscpu output:
```
Architecture:                    aarch64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
CPU(s):                          4
On-line CPU(s) list:             0-3
Thread(s) per core:              1
Core(s) per socket:              4
Socket(s):                       1
Vendor ID:                       ARM
Model:                           3
Model name:                      Cortex-A72
Stepping:                        r0p3
CPU max MHz:                     1800.0000
CPU min MHz:                     600.0000
BogoMIPS:                        108.00
L1d cache:                       128 KiB
L1i cache:                       192 KiB
L2 cache:                        1 MiB
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Vulnerable
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fp asimd evtstrm crc32 cpuid
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @seemethere @malfet @osalpekar @atalman @snadampal @milpuz01

**Comment by malfet** [comment]
Thank you for the report and sorry about regression.
@atalman do we run any pre-release checks on A1? And also, perhaps it's time to consider whether it's OK to retire A72 support

**Comment by atalman** [comment]
hi @malfet no we did not run check on A1 : https://github.com/pytorch/pytorch/issues/130151 This is list of checks that was run for release 2.4.0. 

Added Issue: https://github.com/pytorch/builder/issues/1940 to address adding a1 testing, amazon linux 2023 and stock ubuntu 20.04 testing to validation framework

**Comment by malfet** [comment]
Backtrace
```
$ lldb -- python3 -c "import torch"
(lldb) target create "python3"
Current executable set to 'python3' (aarch64).
(lldb) settings set -- target.run-args  "-c" "import torch"
(lldb) r
Process 8703 launched: '/usr/bin/python3' (aarch64)
Process 8703 stopped and restarted: thread 1 received signal: SIGCHLD
warning: (aarch64) /home/ubuntu/.local/lib/python3.8/site-packages/torch.libs/libgomp-804f19d4.so.1.0.0 unsupported DW_FORM value: 0x1f
warning: (aarch64) /home/ubuntu/.local/lib/python3.8/site-packages/torch.libs/libgfortran-daac5196.so.5.0.0 No LZMA support found for reading .gnu_debugdata section
Process 8703 stopped
* thread #1, name = 'python3', stop reason = signal SIGILL: illegal instruction
    frame #0: 0x0000ffffeb9b3800 libarm_compute-6cddc52f.so`std::string::_Rep::_M_dispose(std::allocator<char> const&) (.part.0) + 44
libarm_compute-6cddc52f.so`std::string::_Rep::_M_dispose(std::allocator<char> const&) (.part.0):
->  0xffffeb9b3800 <+44>: ldaddal w2, w2, [x3]
    0xffffeb9b3804 <+48>: b      0xffffeb9b37ec            ; <+24>
    0xffffeb9b3808 <+52>: b      0xffffeb3fc4f0            ; symbol stub for: std::string::_Rep::_M_destroy(std::allocator<char> const&)
    0xffffeb9b380c:       nop    
(lldb) bt
* thread #1, name = 'python3', stop reason = signal SIGILL: illegal instruction
  * frame #0: 0x0000ffffeb9b3800 libarm_compute-6cddc52f.so`std::string::_Rep::_M_dispose(std::allocator<char> const&) (.part.0) + 44
    frame #1: 0x0000ffffeb41a030 libarm_compute-6cddc52f.so`_GLOBAL__sub_I_weight_transforms_fp16.cpp + 304
```

**Comment by malfet** [comment]
Following patch against ACL fixes the problem:
```
```patch
diff --git a/src/core/NEON/kernels/convolution/winograd/input_transforms_fp16.cpp b/src/core/NEON/kernels/convolution/winograd/input_transforms_fp16.cpp
index 35d61fa94d..fdee43672e 100644
--- a/src/core/NEON/kernels/convolution/winograd/input_transforms_fp16.cpp
+++ b/src/core/NEON/kernels/convolution/winograd/input_transforms_fp16.cpp
@@ -38,14 +38,14 @@ void a64_fp16_6x6(unsigned int, const __fp16 *, size_t, size_t, __fp16 *, size_t
 
 #define IMPL(HEIGHT, WIDTH, FUNC, DRIVER) new Transform ## DRIVER <__fp16, __fp16>(#FUNC, HEIGHT, WIDTH, FUNC)
 
-static const TransformImplementation<__fp16> transforms_fp16[] = {
-  { IMPL(6, 6, a64_fp16_6x6, Unpadded) },
-  { nullptr },
-};
 
 template <>
 const TransformImplementation<__fp16> *implementation_list(void)
 {
+static const TransformImplementation<__fp16> transforms_fp16[] = {
+  { IMPL(6, 6, a64_fp16_6x6, Unpadded) },
+  { nullptr },
+};
   return transforms_fp16;
 }
 
diff --git a/src/core/NEON/kernels/convolution/winograd/output_transforms_fp16.cpp b/src/core/NEON/kernels/convolution/winograd/output_transforms_fp16.cpp
index c39b1dc083..b312952340 100644
--- a/src/core/NEON/kernels/convolution/winograd/output_transforms_fp16.cpp
+++ b/src/core/NEON/kernels/convolution/winograd/output_transforms_fp16.cpp
@@ -37,14 +37,14 @@ void a64_fp16_4x4_3x3(unsigned int, const __fp16 *, size_t, const __fp16 *, __fp
   new Transform ## DRIVER <__fp16, __fp16>(#FUNC, OUT_HEIGHT, OUT_WIDTH, KERN_HEIGHT, KERN_WIDTH, FUNC)
 
 
+template <>
+const TransformImplementation<__fp16> *implementation_list(void)
+{
 static const TransformImplementation<__fp16> transforms_fp16[] = {
   { IMPL(4, 4, 3, 3, a64_fp16_4x4_3x3, Unpadded) },
   { nullptr }
 };
 
-template <>
-const TransformImplementation<__fp16> *implementation_list(void)
-{
   return transforms_fp16;
 }
 
@@ -52,4 +52,4 @@ const TransformImplementation<__fp16> *implementation_list(void)
 }  // namespace winograd
 }  // namespace arm_conv
 
-#endif // defined(__aarch64__) && defined(__ARM_FEATURE_FP16_VECTOR_ARITHMETIC)
\ No newline at end of file
+#endif // defined(__aarch64__) && defined(__ARM_FEATURE_FP16_VECTOR_ARITHMETIC)
diff --git a/src/core/NEON/kernels/convolution/winograd/weight_transforms_fp16.cpp b/src/core/NEON/kernels/convolution/winograd/weight_transforms_fp16.cpp
index 6c8bbe07cf..91f11a4a7b 100644
--- a/src/core/NEON/kernels/convolution/winograd/weight_transforms_fp16.cpp
+++ b/src/core/NEON/kernels/convolution/winograd/weight_transforms_fp16.cpp
@@ -36,14 +36,14 @@ void *a64_fp16_4x4_3x3(unsigned int, const __fp16 *, size_t, size_t, __fp16 *, s
 #define IMPL(KERN_ROWS, KERN_COLS, TRANS_ROWS, TRANS_COLS, KERN) \
   new Transform<__fp16>(#KERN, KERN_ROWS, KERN_COLS, TRANS_ROWS, TRANS_COLS, KERN)
 
-static const TransformImplementation<__fp16> transforms_fp16[] = {
-  { IMPL(3, 3, 6, 6, a64_fp16_4x4_3x3) },
-  { nullptr }
-};
 
 template <>
 const TransformImplementation<__fp16> *implementation_list(void)
 {
+static const TransformImplementation<__fp16> transforms_fp16[] = {
+  { IMPL(3, 3, 6, 6, a64_fp16_4x4_3x3) },
+  { nullptr }
+};
   return transforms_fp16;
 }
```
```

**Comment by malfet** [comment]
Landed https://github.com/pytorch/builder/commit/4a08b9a8cf657512eb9b60dd940be0bb01d7578d that should help, if nightlies are good, will cherry-pick into release/2.4

**Comment by tecandrew** [comment]
hi! I'm seeing this issue as well on another arm64 based PC, but is not a Raspberry Pi

### `lscpu`

```
root@imx8mpsolidrun:~/xx/audio_processing# lscpu 
Architecture:            aarch64
  CPU op-mode(s):        32-bit, 64-bit
  Byte Order:            Little Endian
CPU(s):                  4
  On-line CPU(s) list:   0-3
Vendor ID:               ARM
  Model name:            Cortex-A53
    Model:               4
    Thread(s) per core:  1
    Core(s) per cluster: 4
    Socket(s):           -
    Cluster(s):          1
    Stepping:            r0p4
    CPU max MHz:         1800.0000
    CPU min MHz:         1200.0000
    BogoMIPS:            16.00
    Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid
NUMA:                    
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-3
Vulnerabilities:         
  Itlb multihit:         Not affected
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Mmio stale data:       Not affected
  Retbleed:              Not affected
  Spec store bypass:     Not affected
  Spectre v1:            Mitigation; __user pointer sanitization
  Spectre v2:            Not affected
  Srbds:                 Not affected
  Tsx async abort:       Not affected
 ```
 
torch              2.4.0
torchaudio         2.4.0
torchlibrosa       0.1.0
python 3.10.9

### GDB dump

<details>

```
(audio_processing) root@imx8mpsolidrun:~/xx/audio_processing# gdb python3
GNU gdb (GDB) 11.2
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "aarch64-poky-linux".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python3...
(gdb) start
Temporary breakpoint 1 at 0xa18: file ../Python-3.10.9/Programs/python.c, line 14.
Starting program: /home/root/xx/audio_processing/.venv/bin/python3 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/libthread_db.so.1".

Temporary breakpoint 1, main (argc=1, argv=0xfffffffffb18) at ../Python-3.10.9/Programs/python.c:14
14	../Python-3.10.9/Programs/python.c: No such file or directory.
(gdb) c
Continuing.
Python 3.10.9 (main, Dec  6 2022, 18:44:57) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
warning: Cannot parse .gnu_debugdata section; LZMA support was disabled at compile time

Program received signal SIGILL, Illegal instruction.
0x0000ffffeb195040 in std::string::_Rep::_M_dispose(std::allocator<char> const&) [clone .part.0] () from /home/root/xx/audio_processing/.venv/lib/python3.10/site-packages/torch/lib/../../torch.libs/libarm_compute-d8ad1c91.so
(gdb) 
A debugging session is active.

	Inferior 1 [process 1062] will be killed.

Quit anyway? (y or n) y
```

</details>

**Comment by christinaburge** [comment]
The proposed patch on the issue has been merged into ACL.

**Comment by maxkonrad** [comment]
mentioned here [openwebui/openwebui#4331](https://github.com/open-webui/open-webui/discussions/4331) fyi

**Comment by malfet** [comment]
> hi! I'm seeing this issue as well on another arm64 based PC, but is not a Raspberry Pi
> 
> ### `lscpu`
> ```
>   Model name:            Cortex-A53
>```

Yes, this would affect any Cortex-A53 based system. Change has been cherry-picked into release/2.4 branch and I hope to see RC build with the fix real soon

**Comment by malfet** [comment]
Ok, change has been cherry-picked into the branch and 2.4.1 release candidate (available at https://download.pytorch.org/whl/test/cpu ) is no longer affected by this problem

**Comment by Module32** [comment]
> Ok, change has been cherry-picked into the branch and 2.4.1 release candidate (available at https://download.pytorch.org/whl/test/cpu ) is no longer affected by this problem

Is this available for the ARM-based Raspberry Pis yet?

**Comment by tecandrew** [comment]
confirmed that v2.4.1 fixes issues on both `pi` and `Cortex-A53` machines
<img width="1713" alt="image" src="https://github.com/user-attachments/assets/58ab865d-282f-41bb-9c9d-cee7e4adda37">

**Comment by malfet** [comment]
Thank you, closing as fixed!

**Comment by dsx724** [comment]
@malfet This issue has regressed in PyTorch 2.10.0. Just tested on Cortex-A53 and import torch still crashes with SIGILL in _mi_options_init() at ldaddal instruction. The fix from 2.4.1 appears to have been lost in later releases. Should this be reopened or should I create a new ticket? Seems like a recurring problem.

Mimalloc is still enabling ARMv8.1 LSE on all ARM64 builds. Should runtime detection be added or should build flag be modified?
